pkg/hive/job/job.go

```diff
@@ -0,0 +1,485 @@
+		}
+	}
+
+	if options.shutdowner != nil && jos.shutdownOnError {
+		options.shutdowner.Shutdown(hive.ShutdownWithError(err))
+	}
+
+	return err
+}
+
+// Timer creates a timer job which can be added to a Group. Timer jobs invoke the given function at the specified
+// interval. Timer jobs are particularly useful to implement periodic syncs and cleanup actions.
+// Timer jobs can optionally be triggered by an external Trigger with the WithTrigger option.
+// This trigger can for example be passed between cells or between jobs in the same cell to allow for an additional
+// invocation of the function.
+//
+// The interval between invocations is counted from the start of the last invocation. If the `fn` takes longer than the
+// interval, its next invocation is not delayed. The `fn` is expected to stop as soon as the context passed to it
+// expires. This is especially important for long running functions. The signal created by a Trigger is coalesced so
+// multiple calls to trigger before the invocation takes place can result in just a single invocation.
+func Timer(name string, fn TimerFunc, interval time.Duration, opts ...timerOpt) Job {
+	job := &jobTimer{
+		name:     name,
+		fn:       fn,
+		interval: interval,
+		opts:     opts,
+	}
+
+	return job
+}
+
+// TimerFunc is the func type invoked by a timer job. A TimerFunc is expected to return as soon as the ctx expires.
+type TimerFunc func(ctx context.Context) error
+
+type timerOpt func(*jobTimer)
+
+// NewTrigger creates a new trigger, which can be used to trigger a timer job.
+func NewTrigger() *trigger {
+	return &trigger{
+		c: make(chan struct{}, 1),
+	}
+}
+
+type trigger struct {
+	c chan struct{}
+}
+
+func (t *trigger) Trigger() {
+	select {
+	case t.c <- struct{}{}:
+	default:
+	}
+}
+
+// WithTrigger option allows a user to specify a trigger, which if triggered will invoke the function of a timer
+// before the configured interval has expired.
+func WithTrigger(trigger *trigger) timerOpt {
+	return func(jt *jobTimer) {
+		jt.trigger = trigger
+	}
+}
+
+type jobTimer struct {
+	name string
+	fn   TimerFunc
+	opts []timerOpt
+
+	interval time.Duration
+	trigger  *trigger
+
+	// If not nil, call the shutdowner on error
+	shutdown hive.Shutdowner
+}
+
+func (jt *jobTimer) start(ctx context.Context, wg *sync.WaitGroup, options options) error {
+	defer wg.Done()
+
+	for _, opt := range jt.opts {
+		opt(jt)
+	}
+
+	if jt.fn == nil {
+		return fmt.Errorf("can't start a nil one shot job")
+	}
+
+	l := options.logger
+	if options.logger == nil {
+		l = &logrus.Logger{
+			Out: io.Discard,
+		}
+	}
+
+	fields := logrus.Fields{
+		"name": jt.name,
+		"func": internal.FuncNameAndLocation(jt.fn),
+	}
+
+	timer := time.NewTicker(jt.interval)
+	defer timer.Stop()
+
+	var triggerChan chan struct{}
+	if jt.trigger != nil {
+		triggerChan = jt.trigger.c
+	}
+
+	l.WithFields(fields).Debug("Starting timer job")
+
+	for {
+		select {
+		case <-triggerChan:
+		case <-timer.C:
+		case <-ctx.Done():
```

Comment:
maybe shuffle the cases so that `<-ctx.Done()` is first. For an old C hacker like me I'm still having hard time not reading this as all falling through to `return nil`.